File: csharp/maze_scripts/helper_classes/TileEdge.cs
using Godot;

namespace MazePerformanceGrade.csharp.helper_classes;

public readonly record struct TileEdge
{
    public Vector2I A { get; }
    public Vector2I B { get; }
    
    public TileEdge(Vector2I a, Vector2I b)
    {
        var delta = b - a;
        if(delta.X != 0 && delta.Y != 0)
            throw new System.ArgumentException("edges must be either horizontal or vertical");
        // enforce a consistent ordering. order by x coord first, then ycoord if x is equal
        //  this is to make sure that the same edge is always represented the same way
        if (delta.X != 0)
        {
            if (delta.X < 0) (b, a) = (a, b);
        }
        else
        {
            if(delta.Y < 0) (b, a) = (a, b);
        }
        A = a;
        B = b;
    }

    
    public static Vector3 To3DFrom2D(Vector2I flat)
    {
        return new Vector3(flat.X, 0, flat.Y);
    }

    public int GetStableIdInDomain(Vector2I size)
    {
        var minVec = Utils.PartsWise(A, B, Mathf.Min);
        var maxVec = Utils.PartsWise(A, B, Mathf.Max);
        
        var min = minVec.X + minVec.Y * size.X;
        var max = maxVec.X + maxVec.Y * size.X;
        return min + max * (size.X * size.Y);
    }
    
    public void SpawnWallAtEdge(PackedScene wallPrefab, Node parent)
    {
        var worldA = To3DFrom2D(A);
        var worldB = To3DFrom2D(B);
        
        var center = (worldA + worldB) / 2;
        var forward = (worldB - worldA).Normalized();
        var newWall = wallPrefab.InstantiateForceCast<Node3D>();
        newWall.Transform = new Transform3D(Basis.LookingAt(forward, Vector3.Up), center);
        parent.AddChild(newWall);
    }
}

File: rust/src/maze_scripts/helper_classes/tile_edge.rs
use godot::prelude::*;

#[derive(PartialEq, Eq)]
pub struct TileEdge{
    pub a: Vector2i,
    pub b: Vector2i,
}

impl TileEdge{
    pub fn new(a: Vector2i, b: Vector2i) -> Self {
        let delta = b - a;
        if delta.x == 0 && delta.y == 0 {
            panic!("self-pointing edge!")
        }

        if delta.x != 0 {
            if delta.x < 0 {
                return Self{a:b, b:a};
            }
        }else{
            if delta.y <0 {
                return Self{a:b, b:a};
            }
        }
        Self{a, b}
    }


    pub fn spawn_wall_at_edge(&self, wall_prefab: &Gd<PackedScene>, parent: &mut Gd<Node3D>){
        let world_a = to_3d_from_2d(self.a);
        let world_b = to_3d_from_2d(self.b);

        let center = (world_a + world_b) / 2.0;
        let forward = (world_b - world_a).normalized();
        let mut new_wall = wall_prefab.instantiate_as::<Node3D>();
        new_wall.set_transform(Transform3D::new(
            Basis::new_looking_at(forward, Vector3::UP, false),
            center
        ));
        parent.add_child(new_wall.upcast::<Node>())
    }
}

pub fn to_3d_from_2d(flat: Vector2i) -> Vector3{
    Vector3::new(flat.x as real, 0.0, flat.y as real)
}